Add state.RouteEntry class

cathay4t opened the pull request:
The `state.RouteEntry` class is intend to be used inside `state.py`
for deduplicating the desire route entires and removing
absent route entries.

EdDev commented on libnmstate/state.py:
```diff
@@ -30,6 +30,43 @@
+    def __init__(self, route):
+        self.state = copy.deepcopy(route)
+        if route.get(Route.STATE) != Route.ABSENT:
+            if Route.TABLE_ID not in route:
```
Please call this section by a proper name under its own function.
I guess it is something like: `complement_defaults`.

I suggest you also mention in the commit message the `why` part. It is unclear why you want to do it.
I would even argue that this does not need to be included here if the explanation for it is seen in a later change.

cathay4t replied:
Why do I need a function if that function is only used for one time and it only contains 6 lines?
How hard is that for a developer to understand I am filling the default value if missing.

EdDev replied:
Functions are also used to structure, abstract and describe what is performed (vs how). Based on your reasoning we will end up with a 1000 lines in a function.
The reader should read the code like a story and decide if he wants details or not (what vs how).

EdDev commented on libnmstate/state.py:
```diff
@@ -30,6 +30,43 @@
+                self.state[Route.NEXT_HOP_ADDRESS] = ''
+
+    def __hash__(self):
+        return hash((self.state.get(Route.STATE),
```
I think this shows why saving a dict as the object state is improper here.
You need to re-access the value for each key, on every call to `__hash__`, instead of just directly specifying the attributes (and extracting the values once in the constructor).

You have an example here of something similar: https://github.com/nmstate/nmstate/blob/master/libnmstate/nm/wired.py#L27

cathay4t replied:
The `class State` will use that dict for future handling. 
Saving them as object property will force me to do `to_dict()` in `State.merge_config()`.

cathay4t replied:
If this is the deal breaker, let me know I can do extra work to make this PR merged.

EdDev replied:
I consider the encapsulated data to server the class and not to be transparent to the user, especially when it is "complex". If the inner usage needs to access the properties, it makes it reasonable to expose them as attributes internally and do the translation for the input and output to whatever format is needed.
If you look at this object now, without the usage, you have hard time explaining why you keep it as a dict internally, which is a smell that needs attention.

EdDev commented on libnmstate/state.py:
```diff
@@ -30,6 +30,43 @@
+
+    def is_match(self, other):
+        """
+        Assuming the other state contains all the properties to compare.
```
It will be better to avoid assumptions of what is the `other`.
It is ok to define the expectations from `other` in general: self and other states are assumed to contain all properties.
However, if you use attributes instead of a dict, you will not need this at all, the type assures that.
I also do not understand why we need two equality versions (this one and `__eq__`).

cathay4t replied:
I cannot use `attributes` as absent route allows missing/wildcard property which attributes/properties cannot handle.

EdDev replied:
How is it different from the current implementation in this regard? A None valued attribute is the same as a missing key in the dict. Do I miss something?
I am not sure how you want to use Route that represent an absent, maybe when that will be clearer other things will get clarified.

cathay4t replied:
Treat None as missing property.

cathay4t replied:
Function changed to `can_remove()` and comment changed to:
`Return True if other route should be removed by self absent route.`

EdDev commented on tests/lib/state_test.py:
```diff
@@ -138,3 +140,138 @@ def _extra_state(self):
+
+
+class TestAssertRouteState(object):
+    parametrize_ip_ver_routes = pytest.mark.parametrize(
```
I do not understand how this is relevant to the tested code. The tested code is generic, it does not care what IP version is used, it actually does not even care what is the content.

What is the minimum number of routes you need for these tests? 2? 3?

cathay4t replied:
Changed to function `_get_mixed_test_routes()` which returns 1 IPv4 route and 1 IPv6 route.

EdDev commented on tests/lib/state_test.py:
```diff
@@ -138,3 +140,138 @@ def _extra_state(self):
+
+    @parametrize_ip_ver_routes
+    def test_hash_unique(self, routes):
+        assert hash(state._Route(routes[0])) == hash(state._Route(routes[0]))
```
You still need to check the negative:
`assert hash(state._Route(routes[0])) != hash(state._Route(routes[1]))`

Here is a buggy implementation that this test is not catching:
```
def __hash__(self):
    hash(a, b, c, d, ...)
```

cathay4t replied:
Your understanding of `hash()` is wrong. The hash function can generate the same result for two
different data. So we can never know for sure we can get different hash results for different tuple.

EdDev replied:
Please respond on how you cover the bug example.
Although in theory the negative test may fail, it is unlikely. You can add a comment that warns for it if you like.

EdDev commented on tests/lib/state_test.py:
```diff
@@ -138,3 +140,138 @@ def _extra_state(self):
+    def test_obj_unique(self, routes):
+        route = routes[0]
+        new_route = copy.deepcopy(routes[0])
+        assert state._Route(route) == state._Route(new_route)
```
I'm pretty sure that on py2 this test negative will not pass:
`assert not (state._Route(route) != state._Route(new_route))`

Or just: `assert state._Route(route) != state._Route(different_route)`

EdDev replied:
> Or just: assert state._Route(route) != state._Route(different_route)

Correction, I meant that this will pass: assert state._Route(route) != state._Route(new_route) 
Because it will check for identity instead of equality.

cathay4t replied:
The tox passed for running on py27 about the test negative

cathay4t replied:
negative test cases added.

EdDev commented on libnmstate/state.py:
```diff
@@ -30,6 +30,43 @@
+                     self.state.get(Route.METRIC)))
+
+    def __eq__(self, other):
+        return self.state == other.state
```
Usually an identity check is done in such cases for optimization.

cathay4t replied:
Can you elaborate on this?

cathay4t commented on libnmstate/state.py:
```diff
@@ -30,6 +30,43 @@
+from libnmstate.schema import Route
+
+
+class _Route(object):
```
It's been used inside this python file and not mean to be used outside of `state.py`.
Are you suggesting I should expose it as `class State` did?

EdDev replied:
I’m just telling you the rules of engagement here. Hiding the class will require you to test it through its encapsulator (State I guess) which is probably much harder. In such cases I usually choose to expose it.

cathay4t commented on libnmstate/state.py:
```diff
@@ -30,6 +30,43 @@
+class _Route(object):
+    def __init__(self, route):
+        self.state = copy.deepcopy(route)
+        if route.get(Route.STATE) != Route.ABSENT:
```
This class does not handle removal, the `class State` will handle it.
This is also the con why small PR cannot show the big picture.

EdDev replied:
I am just seeking to understand what it does represent and how it fits with the description in the commit message.
How do you expect it to be used when it represents an absent route?

cathay4t replied:
The class did nothing but save the absent route without any modification.

cathay4t commented:
@EdDev New patch uploaded.

Changes:
 * Renamed the class to `state.RouteEntry`.
 * Changed the class to hold data as property instead of dict.
 * Renamed the `is_match()` to `can_remove()` to reflect its usage.
 * Added the negative test on `RouteEntry` object equality.

tyll commented on libnmstate/state.py:
```diff
@@ -32,6 +32,68 @@
+class RouteEntry(object):
+
+    # Mapping dict key name to RouteEntry property name
+    KEY_PROPETY_MAP = {
```
There is a typo in the constant, an `R` is missing. Correct would be `KEY_PROPERTY_MAP`.

tyll commented on libnmstate/state.py:
```diff
@@ -32,6 +32,68 @@
+    }
+
+    def __init__(self, route):
+        self.table_id = route.get(Route.TABLE_ID)
```
you could use the `KEY_PROPERTY_MAP` with `setattr()` here to avoid having the list of properties twice.

cathay4t replied:
The pylint or flake8 will contain about later I check `self.table_id` with defined in `__init__`.

tyll commented on libnmstate/state.py:
```diff
@@ -32,6 +32,68 @@
+            return False
+
+        for key, prop in six.viewitems(RouteEntry.KEY_PROPETY_MAP):
+            if key == Route.STATE:
```
what if the other route entry has the value `absent` for the `state` property?

cathay4t replied:
Good point. In that case a absent cannot remove another absent.

cathay4t replied:
Test case added also.

cathay4t commented:
New patch uploaded.

Changes:
 * The `RouteEntry.can_remove()` will return False when trying to remove another absent route.
 * Add test case above.
 * Fix typo of `KEY_PROPETY_MAP`.

EdDev commented on tests/lib/state_test.py:
```diff
@@ -138,3 +140,121 @@ def _extra_state(self):
+        route = routes[0]
+        new_route = copy.deepcopy(routes[0])
+        assert state.RouteEntry(route) == state.RouteEntry(new_route)
+        assert state.RouteEntry(route) != state.RouteEntry(routes[1])
```
I think this checks identity and not content.
I suspect this one will probably pass (when it should not):
`assert state.RouteEntry(route) != state.RouteEntry(new_route)`

EdDev commented on libnmstate/state.py:
```diff
@@ -32,6 +32,72 @@
+    def __eq__(self, other):
+        return self is other or self.to_dict() == other.to_dict()
+
+    def can_remove(self, other):
```
You previous version of `match` was much better.
Methods of an object gets or sets things about itself, not about others. You ask yourself if you can remove another? This is not intuitive.

I think you better use `match` or overwrite `__le__` so you can ask: `a < b`.
The question of its being absent or not is separated (you can use a property).
The caller then can say something like `if a.is_absent and a < b: ...`.

tyll replied:
`is_absent` is not intuitive for me since it is more a command to make several other routes absent and does not say that a certain route entry is absent. I was wondering if it should be renamed to `removes()` or so but was not so sure as well.

EdDev replied:
Not sure what you mean. This object represents a route entry, which can be partial (in terms of its properties). How the object is used should not influence it's methods names.
`is_absent` is just a way to reflect what is its status, but it can be `is_state_absent` if it sounds better. The fact that makes it problematic to you (I am just guessing), is that this object is not representative of an actual route (due to the wildcard). You could argue that this is not a good representation, and that we need two object types, one that represents a real route and one that represents a "wildcard" route.

EdDev commented on libnmstate/state.py:
```diff
@@ -32,6 +32,72 @@
+class RouteEntry(object):
+
+    # Mapping dict key name to RouteEntry property name
+    KEY_PROPERTY_MAP = {
```
Using this one and `getattr` all over the object is IMO no different from the previous version.
The need for the conversion is limited to the `to_dict`, I do not understand its value in other cases.

BTW, you could have used something like this to report the dict:
`return {k.replace('_', '-'): v for k, v in vars(self).items}`

EdDev commented on libnmstate/state.py:
```diff
@@ -32,6 +32,69 @@
+        return hash(self.__keys())
+
+    def __keys(self):
+        return (self.state, self.table_id, self.metric, self.destination,
```
Do you need to check the hash and equality on `self.state` ?

EdDev commented on libnmstate/state.py:
```diff
@@ -32,6 +32,69 @@
+            # Absent route cannot match another absent route
+            return False
+
+        for key, value in six.viewitems(vars(self)):
```
In case `state` can be removed from __keys,  you can also use __keys in the match:
```
for s, o in zip(self.__keys(), other.__keys()):
    if s is None:
        continue
    if s != o:
        return False
return True

cathay4t commented:
@EdDev New patch uploaded.

Changes:
 * Removed the state from `__hash__` and `__eq__`.
 * Simplified the `is_match()` function.
